src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,942 @@
+               std::vector<dgl_id_t> *col_idx,
+               std::vector<dgl_id_t> *orig_eids) {
+    if (!map.test(old_id))
+      return;
+
+    auto it = oldv2newv.find(old_id);
+    if (it != oldv2newv.end()) {
+      const dgl_id_t new_id = it->second;
+      col_idx->push_back(new_id);
+      if (orig_eids)
+        orig_eids->push_back(old_eid);
+    }
+  }
+};
+
+std::pair<ImmutableGraph::csr::ptr, IdArray> ImmutableGraph::csr::VertexSubgraph(
+    IdArray vids) const {
+  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
+  const int64_t len = vids->shape[0];
+
+  HashTableChecker def_check(vid_data, len);
+  // check if varr is sorted.
+  CHECK(std::is_sorted(vid_data, vid_data + len)) << "The input vertex list has to be sorted";
+
+  // Collect the non-zero entries in from the original graph.
+  std::vector<dgl_id_t> orig_edge_ids;
+  orig_edge_ids.reserve(len * 50);
+  auto sub_csr = std::make_shared<csr>(len, len * 50);
+  sub_csr->indptr[0] = 0;
+  for (int64_t i = 0; i < len; ++i) {
+    const dgl_id_t oldvid = vid_data[i];
+    CHECK_LT(oldvid, NumVertices()) << "Vertex Id " << oldvid << " isn't in a graph of "
+        << NumVertices() << " vertices";
+    size_t row_start = indptr[oldvid];
+    size_t row_len = indptr[oldvid + 1] - indptr[oldvid];
+    def_check.CollectOnRow(&indices[row_start], &edge_ids[row_start], row_len,
+                           &sub_csr->indices, &orig_edge_ids);
+    sub_csr->indptr[i + 1] = sub_csr->indices.size();
+  }
+
+  // Store the non-zeros in a subgraph with edge attributes of new edge ids.
+  sub_csr->edge_ids.resize(sub_csr->indices.size());
+  for (int64_t i = 0; i < sub_csr->edge_ids.size(); i++)
+    sub_csr->edge_ids[i] = i;
+
+  IdArray rst_eids = IdArray::Empty({static_cast<int64_t>(orig_edge_ids.size())},
+                                    DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  dgl_id_t* eid_data = static_cast<dgl_id_t*>(rst_eids->data);
+  std::copy(orig_edge_ids.begin(), orig_edge_ids.end(), eid_data);
+
+  return std::pair<std::shared_ptr<ImmutableGraph::csr>, IdArray>(sub_csr, rst_eids);
+}
+
+ImmutableGraph::csr::ptr ImmutableGraph::csr::from_edges(std::vector<edge> *edges,
+                                                         int sort_on, int64_t num_nodes) {
+  assert(sort_on == 0 || sort_on == 1);
+  int other_end = sort_on == 1 ? 0 : 1;
+  // TODO(zhengda) we should sort in parallel.
+  struct compare {
+    int sort_on;
+    int other_end;
+    compare(int sort_on, int other_end) {
+      this->sort_on = sort_on;
+      this->other_end = other_end;
+    }
+    bool operator()(const edge &e1, const edge &e2) {
+      if (e1.end_points[sort_on] == e2.end_points[sort_on])
+        return e1.end_points[other_end] < e2.end_points[other_end];
+      else
+        return e1.end_points[sort_on] < e2.end_points[sort_on];
+    }
+  };
+  std::sort(edges->begin(), edges->end(), compare(sort_on, other_end));
```

Comment:
You can put all the buckets in a single vector.

Reply:
i feel radix sort is difficult here. each element is a pair. since radix sort doesn't exist in STL and the general sort usually works well, I'll leave the code as it is.

Reply:
Fair enough.

Reply:
I actually agree with @BarclayII that radix sort is better since all these values are integers and it is easy to compute hash function. I suggest put a TODO here. I'm thinking about doing another PR that tries to enable graph ops on immutable graph too. I could incorporate this change in that PR.